Added - #104: Allow resizing route column

sotcha opened the pull request:
<!-- 
**IMPORTANT: Please do not create a Pull Request without creating an issue first.**
*Any change needs to be discussed before proceeding. Failure to do so may result in the rejection of the pull request.*
 -->

**Description**
<!-- Describe your changes in detail -->

Added _ResizeColumnDirective_ which : 
- Tracks **mousedown** event
- Adds **mousemove** and **mouseup** to document.body in order to calc the new width and apply it (with css of parent element).

The directive added to 
* _src/app/components/environment-logs.component.html_
* _src/app/components/routes-menu/routes-menu.component.html_

Also, tooltips added to these files for the route and environment logs urls.

**Related Issue**

<!-- This project only accepts pull requests related to open issues -->
<!-- If suggesting a new feature or change, please discuss it in an issue first -->
<!-- If fixing a bug, there should be an issue describing it with steps to reproduce -->
<!-- Please link to the issue here: -->
Issue : #107 

**Motivation and Context**

<!-- Why is this change required? What problem does it solve? -->
<!-- If it fixes an open issue, please link to the issue here. -->

**How Has This Been Tested?**

<!-- Please describe in detail how you tested your changes. -->
<!-- Include details of your testing environment, and the tests you ran to -->
<!-- see how your change affects other areas of the code, etc. -->
It is tested manually :
- Start the server 
- Place the mouse to the routes column right edge and drag it.

_ps : I am not sure how we can write tests for this_


**Screenshots (if appropriate):**
![image](https://user-images.githubusercontent.com/1183265/76226075-07f50980-6226-11ea-92de-2bc5974b693f.png)

**Closing issues**
closes #107 
<!-- Put `closes #XXXX` in your comment to auto-close the issue that your PR fixes. -->

255kb commented on src/app/directives/resize-column.directive.ts:
```diff
@@ -0,0 +1,94 @@
+   * - mouseup: mark as non-pressed
+   */
+  private initResizableColumns() {
+    this.renderer.listen('body', 'mousemove', (event) => {
```
Could you instead use `@HostListener`?

sotcha replied:
Well, I can use `@HostListener`, but it will listen continuously. I think we should register-unregister, which is not supported by HostListener, correct ?

255kb replied:
True, didn't think about this. You can keep it like this 👍 
And yes Angular doesn't have conditional HostListeners https://github.com/angular/angular/issues/7626

255kb commented on src/app/directives/resize-column.directive.ts:
```diff
@@ -0,0 +1,94 @@
+      }
+    });
+
+    this.renderer.listen('body', 'mouseup', (event) => {
```
Could you instead use `@HostListener`?

sotcha commented:
> It looks great, and performances seem to be good.
> Two things I would like to do differently:
> 
> * it would be safer to have hard pixel limits for min and max size that you can pass as the directive config/params. Something like 200-500px for the route menu (which is already very big), and 200-400px for the environment logs. Otherwise, layout can be completely broken due to flx/Bootstrap. Anyway, this will need to be tested thoroughly after setting up these hard limits.
> * these settings would need to be save in the settings file. This is not very complicated. Idea is to:
>   
>   1. add the properties to the `Settings` type (routeMenuSize, envLogsSize),
>   2. add some code to "migrate" the settings in the `migrateSettings` method in settings service,
>   3. call the `updateSettings` method as soon as mouse is released,
>   4. after that settings will be saved automatically
> 
> Thank you!


* Well for the hardcoded min-max size, I added params to the directive
` <resize-column [type]="'routeMenu'" [minWidth]="200" [maxWidth]="500"></resize-column>` 
I think that setting a min is ok, but I am not sure if setting max hardcoded is right.

Let's say that a user has a wide screen and resizes mockoon to be wide, I do see the reason to limit ie _routes menu_ to **500px** .

![image](https://user-images.githubusercontent.com/1183265/76714814-e61dea00-6731-11ea-9e90-1167465fa134.png)

This is an example where _routes menu_ can be bigger.

On the other hand, we can see that the UI looks broken when I resized the window. 

![image](https://user-images.githubusercontent.com/1183265/76714941-9ab80b80-6732-11ea-8e18-bf9987c66902.png)

I think that since the user can manually resize the window, this is acceptable.


* Also for the settings, I added 
```
  routeMenuSize: number;
  envLogsSize: number;
```
for saving the current size, do but we want for the min-max ?

255kb commented:
Saving current sizes for both menus is definitely enough. 
Concerning the max size, I understand, but I would like to avoid people being able to "break" the interface. It will surely back fire into GitHub issues or feeling that UI is not always good. 
Maybe we could set a hardcoded minimum like 200px and dynamic max like you did before? But instead of allowing 80% it could be 40% or something like that. 
But allowing resizing to nearly the whole width, which completely breaks the interface is unacceptable I think, from a user perspective. Also, most softwares I know have a min/max size for resizable items.

sotcha commented:
Hi @255kb , 
Well, I found a better idea (after some tests) to set a minimum width of the right area since this is the element that we need to keep it not compressed. 

`<resize-column [type]="'routeMenu'" [minWidth]="150" [minWidthFromRight]="640"></resize-column>`

And I kept the rule with the max width based on body and a factor, in case that there is no _minWidthFromRight_ passed.


Again, I do not think that solves the issue 100% (the UI in some cases breaks somehow), but seems a close enough solution. I would like to hear your opinion.

255kb commented:
Thank you for the modification. 
I am sorry for the late reply, I was quite busy debugging the last version. 
I am reviewing your work right now and may probably ask for some modification. Let me know if you don't have more time to spend on this. I can also take it from here if needed 👍

255kb commented on src/app/services/settings.service.ts:
```diff
@@ -29,7 +31,9 @@ export class SettingsService {
-    truncateRouteName: true
+    truncateRouteName: true,
+    routeMenuSize: undefined,
+    envLogsSize: undefined
```
Please rename to `envLogsMenuSize` or `logsMenuSize` for consistency.

255kb commented on src/app/directives/resize-column.directive.ts:
```diff
@@ -0,0 +1,144 @@
+import {ElementRef, Renderer2, HostListener, Directive, AfterViewInit, Input} from '@angular/core';
+import {SettingsService, SettingsProperties} from '../services/settings.service';
```
please use absolute import `src/...`

255kb commented on src/app/directives/resize-column.directive.ts:
```diff
@@ -0,0 +1,144 @@
+              private settingsService: SettingsService,
+              private store: Store,
+              private renderer: Renderer2) {
+    elementRef.nativeElement.style.cursor = 'col-resize';
```
This should be moved to the styles.scss file https://github.com/mockoon/mockoon/pull/264/files#diff-354873b4b545926f38dc09efd5af2b6dR315

255kb commented on src/app/directives/resize-column.directive.ts:
```diff
@@ -0,0 +1,144 @@
+   * - mouseup: mark as non-pressed
+   */
+  private initResizableColumns() {
+    // this.renderer.listen('body', 'mousemove', (event) => {this.handleMouseMoveEvent(event)});
```
remove comment

sotcha replied:
Sorry for that

255kb replied:
no need to apologize :)

255kb commented on src/app/directives/resize-column.directive.ts:
```diff
@@ -0,0 +1,144 @@
+
+  ngAfterViewInit() {
+    // Init and set size from settings - if that exists
+    const settings = this.store.get('settings');
+    if (typeof settings !== 'undefined') {
+      let width;
+      if (this.type === 'routeMenu') {
+        width = settings.routeMenuSize;
+      } else {
+        width = settings.envLogsSize;
+      }
+
+      if (typeof width !== 'undefined') { // finally update width if needed
+        this.applyWidthCss(width);
+      }
+    }
```
We could eventually have a race condition where settings are not yet loaded. 
Please use something like this instead:

```suggestion
    this.store.select('settings').pipe(
      filter(settings => !!settings),
      first(),
    ).subscribe((settings) => {
      let width;
      if (this.type === 'routeMenu') {
        width = settings.routeMenuSize;
      } else {
        width = settings.envLogsSize;
      }

      if (typeof width !== 'undefined') { // finally update width if needed
        this.applyWidthCss(width);
      }
    });
```

sotcha commented:
@255kb no worries. I can spend time on this (maybe not immediately, but I have).

sotcha commented:
So if we 
- listen window resize 
- maxWidth = document.body.offsetWidth * this.maxPercentageSize / 100

I think there is no point keeping the _minWidthFromRight_ limitation, correct?

255kb commented:
Yes I think so too. Having a maximum viewport % should be enough. 
I did not forget your PR, I will review it in the coming day or two, and I apologize! 🐱

255kb commented:
Thank you @sotcha it's working like a charm! 
One thing that I would like to modify is to invert the route menu and logs menu initial sizes and maxwidth factors. Routes menu is definitely more important and should be able to take more space than environment logs. 

If you agree I can take it from here, do this small change, resolve the conflict and merge.

sotcha commented:
@255kb Yes please continue it. I am glad that I helped with that. :fireworks:

255kb commented:
@sotcha It's merged! I just fixed one small thing and inverted the sizes. 
Thank you again for this great contribution! 🥳
